### PR TITLE
New "pending badge assertions" feature

### PR DESCRIPTION
**Pull request by nqthqn**
todo
 - [x] read and use value from `?include_pending=false`
 - [ ] ~set default value of recipient_type `issue()` to None~
 - [x] refactor `badge_filter()` 
 - [x] add more tests
 - [ ] ~the `is type(query) is dict`~
 - [x] ~run all tests locally~

### CODE REVIEW

File: apps/badgeuser/tests.py [hunk 1]
```diff
@@ -731,8 +731,8 @@ def test_identifiers_serialized_to_correct_fields(self):
 
     def test_verified_recipient_receives_assertion(self):
         url = 'http://example.com'
-        self.first_user.userrecipientidentifier_set.create(identifier=url, verified=True)
-        self.badgeclass.issue(recipient_id=url)
+        self.first_user.userrecipientidentifier_set.create(identifier=url, verified=True, type=UserRecipientIdentifier.IDENTIFIER_TYPE_URL)
+        self.badgeclass.issue(recipient_id=url, recipient_type=UserRecipientIdentifier.IDENTIFIER_TYPE_URL)
```
Comment (nqthqn) [inline apps/badgeuser/tests.py]: Need to audit our use of `self.badgeclass.issue()` and ensure we are always passing in the identifier type.

File: apps/issuer/helpers.py [hunk 2]
```diff
@@ -138,28 +139,31 @@ def badgecheck_options(cls):
...
 
         # distill 3 optional arguments into one query argument
         query = (url, imagefile, assertion)
         query = filter(lambda v: v is not None, query)
         if len(query) != 1:
             raise ValueError("Must provide only 1 of: url, imagefile or assertion_obo")
         query = query[0]
-
+        
         if created_by:
```
Comment (nqthqn) [inline apps/issuer/helpers.py]: When would `created_by` be falsy?

File: apps/issuer/helpers.py [hunk 3]
```diff
@@ -138,28 +139,31 @@ def badgecheck_options(cls):
...
             }
         else:
             badgecheck_recipient_profile = None
 
         try:
+            if type(query) is dict:
+                try:
+                    query = json.dumps(query)
+                except (TypeError, ValueError):
+                    raise ValidationError("Could not parse dict to json")
```
Comment (nqthqn) [inline apps/issuer/helpers.py]: This is messy. I feel like we should handle this elsewhere and not have to deal with this ambiguity here.

File: apps/issuer/models.py [hunk 4]
```diff
@@ -566,9 +566,10 @@ def cached_pathway_elements(self):
     def cached_completion_elements(self):
         return [pce for pce in self.completion_elements.all()]
 
-    def issue(self, recipient_id=None, evidence=None, narrative=None, notify=False, created_by=None, allow_uppercase=False, badgr_app=None, **kwargs):
+    def issue(self, recipient_id=None, evidence=None, narrative=None, notify=False, created_by=None, allow_uppercase=False, badgr_app=None, recipient_type=u'email', **kwargs):
```
Comment (nqthqn) [inline apps/issuer/models.py]: `recipient_type` should probably default to None? It can't be null though...

File: apps/issuer/models.py [hunk 5]
```diff
@@ -771,18 +772,27 @@ def owners(self):
...
+                existing_email = None
+            return not existing_email or not existing_email.verified
         else:
-            return False
+            try:
+                existing_identifier = UserRecipientIdentifier.cached.get(
+                    identifier=self.recipient_identifier)
+            except UserRecipientIdentifier.DoesNotExist:
+                existing_identifier = None
+            return not existing_identifier or not existing_identifier.verified
```
Comment (nqthqn) [inline apps/issuer/models.py]: This could use a refactor.

**Comment by nqthqn** [comment]
Should be !=. Not `is not`.

File: apps/issuer/helpers.py (hunk 3 above)
Comment (ottonomy) [reply apps/issuer/helpers.py]: Agree. TODO: fix that in the future.

**Comment by ottonomy** [comment]
This has been merged into release/charlie and will be merged into develop upon release. Thanks!

**Pull request closed without merging**